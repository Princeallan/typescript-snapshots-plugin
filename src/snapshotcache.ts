import * as ts_module from "typescript/lib/tsserverlibrary";
import * as fs from "fs";
import { SnapshotDefinition, parseSnapshotFile } from "./utils";

interface Cache {
    definitions: SnapshotDefinition[];
    lastModifiedTime: number;
}

/**
 * Simple snapshot source cache
 */
export class SnapshotCache {
    public constructor(
        private ts: typeof ts_module,
        private cacheMap: Map<string, Cache> = new Map()
    ) { }

    /**
     * Return snapshot definitions for given snapshot file name.
     * May return definitions from cache if snapshot file hasn't been changed
     *
     * @param fileName
     * @returns
     */
    public parseSnapshot(fileName: string): SnapshotDefinition[] {
        if (!fs.existsSync(fileName)) {
            return [];
        }
        try {
            const { mtime } = fs.statSync(fileName);
            const cacheEntry = this.cacheMap.get(fileName);
            if (cacheEntry && cacheEntry.lastModifiedTime === mtime.getTime()) {
                return cacheEntry.definitions;
            }
            const file = fs.readFileSync(fileName, "utf8");
            const newEntry: Cache = {
                lastModifiedTime: mtime.getTime(),
                definitions: parseSnapshotFile(this.ts, file)
            };
            this.cacheMap.set(fileName, newEntry);
            return newEntry.definitions;
        } catch {
            return [];
        }
    }
}