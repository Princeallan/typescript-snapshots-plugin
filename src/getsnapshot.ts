import * as ts_module from "typescript/lib/tsserverlibrary"
import * as path from "path";
import { SnapshotDefinition, findNodeAtPosition, isMatchingIdentifier, getParentTestBlocks, getCountOfIdentifiersInBlock } from "./utils";
import { SnapshotCache } from "./snapshotcache";

/**
 * Try to get snapshot definition for given position in source file
 *
 * @export
 * @param ts
 * @param sourceFile
 * @param position
 * @param snapshotCache
 * @param snapshotCallIdentifiers
 * @param testBlockIdentifiers
 * @returns
 */
export function tryGetSnapshotForPosition(
    ts: typeof ts_module,
    sourceFile: ts_module.SourceFile | undefined,
    position: number,
    snapshotCache: SnapshotCache,
    snapshotCallIdentifiers: string[],
    testBlockIdentifiers: string[]
): SnapshotDefinition | undefined {
    if (!sourceFile) {
        return;
    }
    try {
        const node = findNodeAtPosition(ts, sourceFile, position);
        if (node && isMatchingIdentifier(ts, node, snapshotCallIdentifiers) && node.parent && node.parent.parent && ts.isCallExpression(node.parent.parent)) {
            const snapshotPath = getSnapshotPathForFileName(sourceFile.fileName);

            // avoid reading snapshot file until there will be real case for snapshot existing, i.e. blockInfo not undefined
            const blockInfo = getParentTestBlocks(ts, sourceFile, testBlockIdentifiers, node.getStart(sourceFile));
            if (blockInfo) {
                const snapshotBlocks = snapshotCache.parseSnapshot(snapshotPath);
                const snapshotCallsInBlock = getCountOfIdentifiersInBlock(ts, blockInfo.lastNode, snapshotCallIdentifiers, node.getStart(sourceFile));

                const customName = node.parent.parent.arguments[0] && ts.isStringLiteralLike(node.parent.parent.arguments[0]) ? node.parent.parent.arguments[0] : undefined;
                // let snapshotName = blockInfo.blockNames.join(" ") + " " + (snapshotCallsInBlock + 1);
                let snapshotName = blockInfo.blockNames.join(" ");
                if (customName) {
                    snapshotName += ": " + (customName as ts.StringLiteralLike).text + " " + snapshotCallsInBlock.namedCalls[(customName as ts.StringLiteralLike).text];
                } else {
                    snapshotName += " " + (snapshotCallsInBlock.anonymousCalls);
                }
                return snapshotBlocks.find(t => t.name === snapshotName);
            }
        }
    } catch {
        /* ignore */
    }
    return undefined;
}


/**
 * Return snapshot file path for given file path
 *
 * @export
 * @param filePath
 * @returns
 */
export function getSnapshotPathForFileName(filePath: string): string {
    return path.dirname(filePath) + "/__snapshots__/" + path.basename(filePath) + ".snap";
}