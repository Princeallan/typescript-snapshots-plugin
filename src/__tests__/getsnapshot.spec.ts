import * as path from "path";
import * as ts from "typescript";
import { getSnapshotPathForFileName, tryGetSnapshotForPosition } from "../getsnapshot";
import { SnapshotCache } from "../snapshotcache";
import { source } from "./testsource";


describe("getSnapshotPathForFileName", () => {
    it("Returns snapshot path for file name", () => {
        expect(getSnapshotPathForFileName("/abc/def/test.ts")).toBe("/abc/def/__snapshots__/test.ts.snap");
    });
})

describe("tryGetSnapshotForPosition", () => {
    const cache = new SnapshotCache(ts as any);
    const file = ts.createSourceFile(path.resolve(__dirname, "./testsource.ts"), source, ts.ScriptTarget.Latest, true, ts.ScriptKind.TS);
    const testIdentifiers = [
        "it", "it.only", "describe"
    ];
    const snapshotIdentifiers = ["toMatchSnapshot"];

    it("Returns undefined for invalid position", () => {
        expect(tryGetSnapshotForPosition(ts as any, file, 2000, cache, snapshotIdentifiers, testIdentifiers)).toBeUndefined();
    });

    it("Returns undefined if position is not at snapshot identifier call", () => {
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 4, 0), cache, snapshotIdentifiers, testIdentifiers)).toBeUndefined();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 5, 6), cache, snapshotIdentifiers, testIdentifiers)).toBeUndefined();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 20, 16), cache, snapshotIdentifiers, testIdentifiers)).toBeUndefined();
    });

    it("Returns undefined for invalid node", () => {

        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 13, 21), cache, snapshotIdentifiers, testIdentifiers)).toBeUndefined();
    });

    it("Returns snapshot definition", () => {
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 2, 18), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 6, 19), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 11, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 25, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 26, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 27, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 32, 33), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 36, 33), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();

        // custom names
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 43, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 44, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 45, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 46, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 47, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 48, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 49, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();
        expect(tryGetSnapshotForPosition(ts as any, file, ts.getPositionOfLineAndCharacter(file, 50, 21), cache, snapshotIdentifiers, testIdentifiers)).toMatchSnapshot();

    });
});