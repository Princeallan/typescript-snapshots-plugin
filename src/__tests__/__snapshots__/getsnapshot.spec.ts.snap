// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tryGetSnapshotForPosition Returns snapshot definition 1`] = `
Object {
  "length": 38,
  "name": "test 1",
  "position": 0,
  "snapshot": "
    first test
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 2`] = `
Object {
  "length": 40,
  "name": "test2 1",
  "position": 41,
  "snapshot": "
    second test
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 3`] = `
Object {
  "length": 44,
  "name": "a test2 1",
  "position": 84,
  "snapshot": "
    a test2 first
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 4`] = `
Object {
  "length": 51,
  "name": "valid test1 1",
  "position": 131,
  "snapshot": "
    valid test first
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 5`] = `
Object {
  "length": 52,
  "name": "valid test1 2",
  "position": 185,
  "snapshot": "
    valid test second
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 6`] = `
Object {
  "length": 51,
  "name": "valid test1 3",
  "position": 240,
  "snapshot": "
    valid test third
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 7`] = `
Object {
  "length": 64,
  "name": "valid inner test2 1",
  "position": 294,
  "snapshot": "
    valid inner test2 first
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 8`] = `
Object {
  "length": 64,
  "name": "valid inner test3 1",
  "position": 361,
  "snapshot": "
    valid inner test3 first
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 9`] = `
Object {
  "length": 66,
  "name": "custom with custom names 1",
  "position": 428,
  "snapshot": "
    custom anonymous 1
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 10`] = `
Object {
  "length": 71,
  "name": "custom with custom names: custom 1",
  "position": 635,
  "snapshot": "
    custom custom 1
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 11`] = `
Object {
  "length": 71,
  "name": "custom with custom names: custom 2",
  "position": 709,
  "snapshot": "
    custom custom 2
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 12`] = `
Object {
  "length": 66,
  "name": "custom with custom names 2",
  "position": 497,
  "snapshot": "
    custom anonymous 2
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 13`] = `
Object {
  "length": 79,
  "name": "custom with custom names: custom2 1",
  "position": 857,
  "snapshot": "
    custom custom second 1
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 14`] = `
Object {
  "length": 66,
  "name": "custom with custom names 3",
  "position": 566,
  "snapshot": "
    custom anonymous 3
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 15`] = `
Object {
  "length": 71,
  "name": "custom with custom names: custom 3",
  "position": 783,
  "snapshot": "
    custom custom 3
",
}
`;

exports[`tryGetSnapshotForPosition Returns snapshot definition 16`] = `
Object {
  "length": 79,
  "name": "custom with custom names: custom2 2",
  "position": 939,
  "snapshot": "
    custom custom second 2
",
}
`;
